### DISCUSSION TITLE
Why are the Form.OnClosed and OnClosing methods not marked with the ObsoleteAttribute?

### DISCUSSION THREAD

**Discussion by geniuszxy** [dotnet/winforms#11720] [Q&A] [answered]
When I want to override the `OnClosed/OnClosing` method in a form class, I always can't distinguish between using `OnClosed/OnClosing` or `OnFormClosed/OnFormClosing` until I read the documentations on MSDN. Since they have been obsolete since .net 2.0, why not add the ObsoleteAttribute  to them?

**Comment by elachlan** [comment, marked as answer]
Can you raise an issue for this? I imagine we just have to raise a PR because someone forgot.

**Reply by paul1956** [reply]
CloseReason was change to be passed to both functions. I have checked several applications and they handle all the clos(ed/ing) functions but the code on the paths is not identical, they handle being called in either order but expect all to be called. Before removing the old functions a real definition of the differences especially when they are called (and that CancelrReason is identical) should be documented.
